### DISCUSSION TITLE
How do I use a different fixture implementation based on a command line argument?

### DISCUSSION THREAD

**Discussion by oakkitten** [pytest-dev/pytest#9828] [Q&A]
I'm testing an application and most test use, directly or indirectly, a fixture called `session_with_profile_loaded`. Normally, I want to tear down the profile after each test, but that can be slow; so I want to have an option to run tests faster—keeping the profile and cleaning up the changes within it. So what I have now is approximately this:

```python
if FAST_BUT_FRAGILE:
    @pytest.fixture
    def session_with_profile_loaded(session_scope_session_with_profile_loaded):
        with profile_state_preserved():
            yield session_scope_session_with_profile_loaded
else:
    @pytest.fixture
    def session_with_profile_loaded(session_scope_empty_session):
        with session_scope_empty_session.profile_loaded():
            yield session_scope_empty_session
```

This works fine, but I want to have `FAST_BUT_FRAGILE` as a command line argument. I don't think I can just have an `if` inside the fixture, as the two implementations depend on different fixtures. And I don't think I can access command line arguments at the time when the fixture is defined. So what's a good way of doing this?

**Comment by symonk** [comment]
Hi @oakkitten on mobile at the minute but you could add a new config option via a `pytest_addoption` hook and use the `request` fixture inside your own fixture to retrieve the command line value if it was provided and act upon, once I get to work I will share a proper example.

**Reply by The-Compiler** [reply]
Note that:

> I don't think I can just have an `if` inside the fixture, as the two implementations depend on different fixtures.

However, I think that approach could be combined with [`request.getfixturevalue()`](https://docs.pytest.org/en/7.1.x/reference.html#pytest.FixtureRequest.getfixturevalue) to write a fixture which checks the commandline argument and then requests one of the other two.

**Reply by RonnyPfannschmidt** [reply]
another mechanism would be to simply register a plugin object with a different fixture implementation in a pytest_configure hook

**Reply by oakkitten** [reply]
thanks! i tried both ways and both work flawlessly. here's option one:

```py
@pytest.fixture
def session_with_profile_loaded(session_scope_empty_session, request):
    if request.config.option.tear_down_profile_after_each_test:
        with session_scope_empty_session.profile_loaded():
            yield session_scope_empty_session
    else:
        session = request.getfixturevalue(
            session_scope_session_with_profile_loaded.__name__
        )
        with profile_state_preserved():
            yield session
```

and here's option two:

```py
def pytest_configure(config):
    class Plugin:
        if config.option.tear_down_profile_after_each_test:
            @pytest.fixture
            def session_with_profile_loaded(self, session_scope_empty_session):
                with session_scope_empty_session.profile_loaded():
                    yield session_scope_empty_session
        else:
            @pytest.fixture
            def session_with_profile_loaded(self, session_scope_session_with_profile_loaded):
                with profile_state_preserved():
                    yield session_scope_session_with_profile_loaded

    config.pluginmanager.register(Plugin())
```

while option two gives a better `--setup-plan`, i think i'll go with the option one as option two makes Pycharm think there's no fixture `session_with_profile_loaded`. also no docstring juggling
